File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void backtrack(vector<int>& nums, vector<int> &list, vector<bool> &vis, vector<vector<int>>&ans){
        if(list.size() == nums.size()){
            ans.push_back(list);
            return;
        }
        for(int i=0; i<nums.size(); i++){
            if(vis[i] == true) continue;
            vis[i] = true;
            list.push_back(nums[i]);
            backtrack(nums,list,vis,ans);
            vis[i] = false;
            list.pop_back();
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<bool>vis(nums.size(), false);
        vector<int>list;
        backtrack(nums,list,vis,result);
        return result;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    List<List<Integer>>ans;
    void helper(int[] nums, int index, List<Integer>list){
        int n = nums.length;
        if(index>=n){
            ans.add(new ArrayList(list));
            return;
        }
        for(int i=0; i<n; i++){
            if(list.contains(nums[i])) continue;
            list.add(nums[i]);
            helper(nums, index+1, list);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        ans = new ArrayList<>();
        helper(nums, 0, new ArrayList<>());
        return ans;
    }
}